pkg/storage/tsdb/bucketindex/loader.go

```diff
@@ -0,0 +1,253 @@
+// index if available, or load it from the bucket otherwise.
+func (l *Loader) GetIndex(ctx context.Context, userID string) (*Index, error) {
+	l.indexesMx.RLock()
+	if entry := l.indexes[userID]; entry != nil {
+		idx := entry.index
+		err := entry.err
+		l.indexesMx.RUnlock()
+
+		// We don't check if the index is stale because it's the responsibility
+		// of the background job to keep it updated.
+		entry.requestedAt.Store(time.Now().Unix())
+		return idx, err
+	}
+	l.indexesMx.RUnlock()
+
+	startTime := time.Now()
+	l.loadAttempts.Inc()
+	idx, err := ReadIndex(ctx, l.bkt, userID, l.logger)
+	if err != nil {
+		// Cache the error, to avoid hammering the object store in case of persistent issues
+		// (eg. corrupted bucket index or not existing).
+		l.cacheIndex(userID, nil, err)
+
+		l.loadFailures.Inc()
+		if !errors.Is(err, ErrIndexNotFound) {
+			level.Error(l.logger).Log("msg", "unable to load bucket index", "user", userID, "err", err)
+		}
+
+		return nil, err
+	}
+
+	// Cache the index.
+	l.cacheIndex(userID, idx, nil)
+
+	elapsedTime := time.Since(startTime)
+	l.loadDuration.Observe(elapsedTime.Seconds())
+	level.Info(l.logger).Log("msg", "loaded bucket index", "user", userID, "duration", elapsedTime)
+	return idx, nil
+}
+
+func (l *Loader) cacheIndex(userID string, idx *Index, err error) {
+	l.indexesMx.Lock()
+	defer l.indexesMx.Unlock()
+
+	// Not an issue if, due to concurrency, another index was already cached
+	// and we overwrite it: last will win.
+	l.indexes[userID] = newCachedIndex(idx, err)
+}
+
+// updateCachedIndexes checks all cached indexes and, for each of them, does two things:
+// 1. Offload indexes not requested since >= idle timeout
+// 2. Update indexes which have been updated last time since >= update timeout
+func (l *Loader) updateCachedIndexes(ctx context.Context) error {
+	var (
+		toUpdate []string
+		toDelete []string
+	)
+
+	// Build a list of users for which we should update or delete the index.
+	l.indexesMx.RLock()
+	for userID, entry := range l.indexes {
+		if time.Since(entry.getRequestedAt()) >= l.cfg.IdleTimeout {
+			toDelete = append(toDelete, userID)
+		} else if entry.err != nil && time.Since(entry.getUpdatedAt()) >= l.cfg.UpdateOnErrorInterval {
+			toUpdate = append(toUpdate, userID)
+		} else if entry.err == nil && time.Since(entry.getUpdatedAt()) >= l.cfg.UpdateOnStaleInterval {
+			toUpdate = append(toUpdate, userID)
+		}
+	}
+	l.indexesMx.RUnlock()
+
+	// Delete unused indexes, if confirmed they're still unused.
+	l.indexesMx.Lock()
+	for _, userID := range toDelete {
+		if idx := l.indexes[userID]; time.Since(idx.getRequestedAt()) >= l.cfg.IdleTimeout {
+			delete(l.indexes, userID)
+			level.Info(l.logger).Log("msg", "unloaded bucket index", "user", userID, "reason", "idle")
+		}
+	}
+	l.indexesMx.Unlock()
+
+	// Update actively used indexes.
+	for _, userID := range toUpdate {
+		readCtx, cancel := context.WithTimeout(ctx, readIndexTimeout)
```

Comment:
To avoid linters panicking due to defer in the loop, we could move entire for body into a method. It would also make `defer` more efficient.